SC development best practices

annenkov opened the pull request:
## Purpose

A section on best practices of SC development.

dg-concordium commented:
@annenkov Will there be more best practices coming? You have an index file so I thought maybe there will be more in future. Otherwise, I would probably get rid of the index file for best practices and just insert the file into the TOC.

annenkov commented:
> @annenkov Will there be more best practices coming? You have an index file so I thought maybe there will be more in future. Otherwise, I would probably get rid of the index file for best practices and just insert the file into the TOC.

@dg-concordium Yes, I thought of it as a folder where we could put additional material, e.g. detailed descriptions of certain security issues, testing and verification techniques, audit best practices, etc.

Bargsteen commented on source/mainnet/net/index.rst:
```diff
@@ -72,6 +72,7 @@ index
    :caption: Smart Contracts v1
 
    Introduction <../smart-contracts/general/introduction>
+   Best practices <../smart-contracts/best-practices/index>
```
I don't think this should be right after the introduction. Perhaps after the on-chain guides but before Ethereum?

Bargsteen commented on source/mainnet/smart-contracts/best-practices/index.rst:
```diff
@@ -0,0 +1,12 @@
+Best practices
+==============
+
+Here, you can find best practices for smart contract development, audit, information about common pitfalls and security vulnerabilities, and how to avoid them.
```
I think we should remove this page for now. Then we can add it once we have more than one page of best practices. I don't see much benefit in having it now, and it just adds an extra click for the user.

annenkov replied:
I wanted to create a structure that could be filled-in with new content, and possibly write some short guidelines on how to contribute. So I could drive the process (and contribute myself), but also encourage contributions by others.
We can discuss how to organise the process on one of the stand-ups.

Bargsteen replied:
Ok, let's keep the structure as is then.

Bargsteen commented on source/mainnet/smart-contracts/best-practices/development.rst:
```diff
@@ -0,0 +1,203 @@
+- the cost of mistakes is very high;
+- possibilities for fixing bugs are limited;
+- the area is evolving constantly, with new vulnerabilities being discovered regularly;
+- malicious parties deliberately try to break your smart contracts and steal the money.
```
I don't know if "steal the money" is too specific. There are types of hacks that don't involve money directly.

annenkov replied:
True, but stealing funds are one of the common goals (and probably the main motivation).
Maybe, "try to steal funds or break your contract in some other way"?

Bargsteen replied:
Yea, just phrase it as an example of things that can occur.

Bargsteen commented on source/mainnet/smart-contracts/best-practices/development.rst:
```diff
@@ -0,0 +1,203 @@
+  The more complexity you have in your contract, the larger the attack surface.
+- Be ready if things go wrong after deployment:
+
+  - provide “pause” functionality;
```
We should also mention the native upgrades here.

Bargsteen commented on source/mainnet/smart-contracts/best-practices/development.rst:
```diff
@@ -0,0 +1,203 @@
+---------------------
+
+- Use ``cargo init`` with an appropriate template to start a new project.
+- Build contract logic as the state struct implementation.
```
Is this what we recommend?

annenkov replied:
I've talked to @abizjak about this at some point, and my understanding was that this is a recommended structure. Most of our non-trivial examples follow this approach. But I might be wrong about this :) So, I'm open to other suggestions.

Bargsteen replied:
I mean it is fine and we do use it. I was just unaware that it was now the recommended approach ;)

Bargsteen commented on source/mainnet/smart-contracts/best-practices/development.rst:
```diff
@@ -0,0 +1,203 @@
+Calling another contract gives control to potentially malicious code that could make arbitrary calls to any other contract, including your own contract.
+Calls to your contract might change its state through entrypoints that permit updating the state, see :ref:`best-practices-reentrancy`.
+Moreover, you should not make any assumptions about energy consumption, or expect that the execution succeeds.
+You contract should be able to correctly handle situations when the call to an external contract fails.
```
```suggestion
Your contract should be able to correctly handle situations when the call to an external contract fails.
```

Bargsteen commented on source/mainnet/smart-contracts/best-practices/development.rst:
```diff
@@ -0,0 +1,203 @@
+^^^^^^^^^^^^^^^^^^^^^^^
+
+- *Avoid complex interactions*. Avoid splitting the on-chain part of your dApp into several smart contracts unless it is strictly necessary.
+  For example, instead of using the *proxy* pattern, use :ref:`natively upgradable contracts <contract-instance-upgradeability>`.
```
Please add a link or more description to "the proxy pattern". People might not be aware of this pattern.

annenkov replied:
I'm not sure where to refer. We have an example, but it's in a pull request. Should I reference some Solidity proxy pattern then? Maybe we can just assume if people don't know, then they don't need to know about it anyways. :)

Bargsteen replied:
I don't like referring to something that users perhaps might not know. We could link to a solidity pattern, mention that it makes everything more complex, and then say that on the superior Concordium blockchain, we can use native upgradability.

Bargsteen commented on source/mainnet/smart-contracts/best-practices/development.rst:
```diff
@@ -0,0 +1,203 @@
+Instead, think of them as sending a message and temporarily pausing execution of your contract.
+The receiving side has full control of what to do next and can choose to call your contract again while it is still in the "paused" state waiting for the external call to be completed.
+Once the external call is completed, you might find your contract in a completely different state.
+See an :ref:`example <reentracny-unit-testing>`, based on the DAO contract vulnerability of how reentrancy can be discovered using unit testing.
```
Please add a link or description to the DAO contract vulnerability.

Bargsteen commented on source/mainnet/smart-contracts/best-practices/development.rst:
```diff
@@ -0,0 +1,203 @@
+- Avoid changing the state after an external call: use the *Checks-Effects-Interactions* pattern: validate data, update the contract state, make external calls.
+- If you need to perform some state changes after an external call use `invoke_contract_read_only <https://docs.rs/concordium-std/latest/concordium_std/trait.HasHost.html#method.invoke_contract_read_only>`_.
+  If the read-only invocation succeeds, it ensures that the state has not been changed after returning from the external call.
+  Using ``invoke_contract`` covers most of the cases where protection from reentrancy is required.
```
Did you mean `invoke_contract_read_only` here ?

annenkov replied:
Yes! Good catch, thanks! ;)

Bargsteen commented on source/mainnet/smart-contracts/best-practices/development.rst:
```diff
@@ -0,0 +1,203 @@
+- If you need to perform some state changes after an external call use `invoke_contract_read_only <https://docs.rs/concordium-std/latest/concordium_std/trait.HasHost.html#method.invoke_contract_read_only>`_.
+  If the read-only invocation succeeds, it ensures that the state has not been changed after returning from the external call.
+  Using ``invoke_contract`` covers most of the cases where protection from reentrancy is required.
+- Alternatively, consider using a *mutex*: a boolean flag that is set before making an external call, preventing all entrypoints from reentrancy. Reset after the call is complete.
```
Perhaps we could include a few lines of code to illustrate this?

Bargsteen commented on source/mainnet/smart-contracts/best-practices/development.rst:
```diff
@@ -0,0 +1,203 @@
+  - What functionality the entrypoint implements?
+  - Who has access rights to call the entrypoint?
+  - When the call is rejected?
+  - What events are logged?
```
and when?

Bargsteen commented on source/mainnet/smart-contracts/best-practices/development.rst:
```diff
@@ -0,0 +1,203 @@
+- :ref:`Unit testing <unit-test-contract>` used for testing particular cases where you define what is the valid output.
+- :ref:`Property-based testing <writing_property_based_tests>` is a variant of randomized testing that repeatedly checks a *property* with randomly generated input.
+
+Use the :ref:`smart contract specification <best-practices-specification>` to come up with cases and properties to test.
```
It wasn't clear to me that this referred to a section in this page.

annenkov replied:
I don't fully understand what you mean here. Do you mean there is another way of referencing the content on the same page? Could you please clarify?

Bargsteen replied:
When I read it, I thought it referred to some other page or guideline and then I was briefly confused about it leading to a section on the same page.

I would just add a few more words, e.g.
```suggestion
Use the :ref:`smart contract specification <best-practices-specification>` guidelines from the section above to come up with cases and properties to test.
```

Bargsteen reviewed (changes requested):
Looks good. Had a few comments 😊

annenkov commented on source/mainnet/smart-contracts/best-practices/development.rst:
```diff
@@ -0,0 +1,203 @@
+- *Avoid complex interactions*. Avoid splitting the on-chain part of your dApp into several smart contracts unless it is strictly necessary.
+  For example, instead of using the *proxy* pattern, use :ref:`natively upgradable contracts <contract-instance-upgradeability>`.
+- *Think about the contract state*. Do not assume that the contract state stays the same after an external call (see :ref:`best-practices-reentrancy` for details).
+- *Protect from denial-of-service (DoS) attacks*. Use the *Pull over Push* pattern: avoid sending funds back (*Push*) to an unknown address as part of some complex operation in your smart contract.
```
Yeah, I agree, there should be a whole section about DoS vulnerabilities :) I can mention these issues without much detail for now.

abizjak commented on source/mainnet/smart-contracts/best-practices/development.rst:
```diff
@@ -0,0 +1,203 @@
+- *Avoid complex interactions*. Avoid splitting the on-chain part of your dApp into several smart contracts unless it is strictly necessary.
+  For example, instead of using the *proxy* pattern, use :ref:`natively upgradable contracts <contract-instance-upgradeability>`.
+- *Think about the contract state*. Do not assume that the contract state stays the same after an external call (see :ref:`best-practices-reentrancy` for details).
+- *Protect from denial-of-service (DoS) attacks*. Use the *Pull over Push* pattern: avoid sending funds back (*Push*) to an unknown address as part of some complex operation in your smart contract.
+  It opens doors to DoS attacks since the contract you call might fail for various reasons, blocking your functionality from succeeding.
+  Instead, create a separate entrypoint allowing users, which could be smart contracts, to request funds back (*Pull*).
```
Yes, I understand what you are referring to. I just think the way that it is explained now makes something that is nuanced, and different approaches have tradeoffs, into a one-sided recommendation.

I also don't understand how the "push pull" terminology is enlightening. To me it does not say much.

annenkov replied:
> I just think the way that it is explained now makes something that is nuanced, and different approaches have tradeoffs, into a one-sided recommendation.

Fair enough, I'll reformulate it.

> I also don't understand how the "push pull" terminology is enlightening. To me it does not say much.

I agree, but this "pull-over-push" is something that is used often, it's not a "standard", but at least something people can recognise, so I think it makes sense to use the same terminology.

annenkov commented:
@Bargsteen I've addressed the comments. The one remaining is whether we should keep the index page referring to one page or not. My point is that it creates some structure for further contributions, so we just add stuff without changing the structure. But I don't know when we're going to have more pages there :)

Bargsteen commented on source/mainnet/smart-contracts/best-practices/development.rst:
```diff
@@ -0,0 +1,295 @@
+  The more complexity you have in your contract, the larger the attack surface.
+- Be ready if things go wrong after deployment:
+
+  - provide “pause” functionality;
```
We mention pausing several times without really describing it. Could you add a note about it or link to some example?

dg-concordium replied:
I believe an example of pausing also addresses another issue in the documentation backlog: https://github.com/Concordium/concordium.github.io/issues/667

Bargsteen replied:
No, that is a different thing enterily. The `interrupt`/`resume` events show up when a contract is invoking certain host functions. For example when it invokes another contract. When that happens, the caller is interrupted, then the callee is invoked (which may also produce events), and finally the caller is `resumed`.

annenkov replied:
I've added a code snippet and pointed to the wCCD example

Bargsteen commented on source/mainnet/smart-contracts/best-practices/development.rst:
```diff
@@ -0,0 +1,295 @@
+        host: &mut impl HasHost<State<S>, StateApiType = S>,
+    ) -> ReceiveResult<()> {
+        // Parse parameters
+        let param : MyParameter = ctx.parameter_cursor().get()?;
```
```suggestion
        let param: MyParameter = ctx.parameter_cursor().get()?;
```

Bargsteen commented on source/mainnet/smart-contracts/best-practices/development.rst:
```diff
@@ -0,0 +1,295 @@
+
+Consider splitting withdrawal of funds from the rest of the contract logic.
+You could create a separate entrypoint allowing users, which could be smart contracts, to request funds back.
+This pattern is called *Pull over Push*, where *Pull* corresponds to user explicitly requesting funds and *Push* to sending the funds back as part of some other operation.
```
```suggestion
This pattern is called *Pull over Push*, where *Pull* corresponds to the user explicitly requesting funds and *Push* to sending the funds back as part of some other operation.
```

Bargsteen commented on source/mainnet/smart-contracts/best-practices/development.rst:
```diff
@@ -0,0 +1,295 @@
+
+This situation is not necessarily an attack, it could occur naturally during the contract lifetime.
+In general, all computations requiring iteration with no clear bound could be an issue.
+For the auction, consider adding ``previous_bid: Amount`` to the state.
+Update ``previous_bid`` once a new bidder proposes a higher bid.
```
```suggestion
For the auction, consider adding ``highest_bid: Amount`` to the state.
Update ``highest_bid`` once a new bidder proposes a higher bid.
```

Bargsteen commented on source/mainnet/smart-contracts/best-practices/development.rst:
```diff
@@ -0,0 +1,295 @@
+
+.. note::
+
+  Keep only relevant data in the contract state.
+  For example, if you are interested in historical data for all the bids, but your contract logic requires only the highest bid to make a decision, consider :ref:`logging events <contract-instances-logging-events>` instead.
+  An off-chain part of your dApp can then use logs to obtain the historical data.
```
Nicely explained 👌

annenkov replied:
Thanks :)

Bargsteen commented on source/mainnet/smart-contracts/best-practices/development.rst:
```diff
@@ -0,0 +1,295 @@
+- Document decisions/choices in the code.
+- Document entrypoints:
+
+  - What functionality the entrypoint implements?
```
```suggestion
  - What functionality does the entrypoint implement?
```

Bargsteen reviewed (approved):
Great work 👍

dg-concordium reviewed (approved):
I made a few minor grammar and readability changes.

DOBEN commented on source/mainnet/smart-contracts/best-practices/development.rst:
```diff
@@ -0,0 +1,336 @@
+
+- What is the interface (entrypoints and their parameters)? For example, ``transfer`` takes three parameters: ``from``, ``to`` addresses, and ``amount``.
+- Who can access the entrypoints? For example, the transaction sender must be an operator of the ``from`` account.
+- Which properties of these entrypoints should be satisfied? For example, a successful call of ``transfer(from, to, amount)`` decreases the ``from`` balance and the ``to`` balance by ``amount``.
```
the ``from`` balance and increases the ``to`` balance by ``amount``.

DOBEN commented on source/mainnet/smart-contracts/best-practices/development.rst:
```diff
@@ -0,0 +1,336 @@
+Recommended structure
+---------------------
+
+- Use ``cargo init`` with an appropriate template to start a new project.
```
``cargo concordium init``

DOBEN commented on source/mainnet/smart-contracts/best-practices/development.rst:
```diff
@@ -0,0 +1,336 @@
+        // Parse parameters
+        let param: MyParameter = ctx.parameter_cursor().get()?;
+        ...
+        // Perform authorization, potentially using using `ctx` info
```
`using` is there twice.

DOBEN reviewed (commented):
A bit late but here are some things that I observed:
